Inclusion of CDXWRF module

LluisFB opened the pull request:
Enhance model output to include CORDEX output demands and new potential useful diagnostics

TYPE: new feature

KEYWORDS: CORDEX, diagnostics, post-process, impact studies, scheme free diagnostics

SOURCE: Lluís Fita Centro de Investigaciones del Mar y la Atmósfera (CIMA), IRL IFAECI UBA-CONICET-CNRS-IRD, Argentina

DESCRIPTION OF CHANGES:
Problem:
The CORDEX-WRF module was initially dedicated to provide the outputs [requested](https://github.com/WCRP-CORDEX/data-request-table) in any [CORDEX](https://cordex.org/) exercise. This tends to be a tedious post-processing work, in sometimes is as long as the simulation by it self. the goal is to provide the diagnostics whilst the integration of the simulation. Enormously reducing the need for post-processing by increasing a little the integration time of the model.

Solution:
A new module has been created which adds about 150 new diagnostics to the model output.
Aside to provide an answer to an specific climate related demand from the CORDEX WRF community, it grew fast to provide:
- Additional non-specific CORDEX diagnostics
- Generic definitions of certain diagnostics that might be scheme dependent
- Introduction of additional variables in the integration-time p-level vertical interpolation

ISSUE: For use when this PR closes an issue.
Fixes #2243 

LIST OF MODIFIED FILES: list of changed files (use `git diff --name-status master` to get formatted list)
```
A       README.cordex
M       README.md
M       Registry/Registry.EM_COMMON
M       Registry/Registry.EM_COMMON.var
A       Registry/registry.cordex
M       Registry/registry.diags
M       Registry/registry.em_shared_collection
M       dyn_em/solve_em.F
M       dyn_em/start_em.F
M       frame/module_driver_constants.F
M       main/depend.common
M       phys/Makefile
A       phys/module_diag_cordex.F
M       phys/module_diag_pld.F
M       phys/module_diagnostics_driver.F
A       phys/module_diagvar_cordex.F
M       phys/module_physics_init.F
```

TESTS CONDUCTED: 
1. Do mods fix problem? How can that be demonstrated, and was that test conducted?
2. Are the Jenkins tests all passing?
Independent tests have been performed aimed to show the correct performance of the module

RELEASE NOTE: 
First version (v1.3) of the module was accompanied by an article in GMD:
Lluís Fita, Jan Polcher, Theodore M. Giannaros, Torge Lorenz, Josipa Milovac, Giannis Sofiadis, Eleni Katragkou and Sophie Bastin, 2019: CORDEX-WRF v1.3: development of a module for the Weather Research and Forecasting (WRF) model to support the CORDEX community, Geosci. Model Dev., 12, 1029-1066, 2019, doi [10.5194/gmd-12-1029-2019](https://www.geosci-model-dev.net/12/1029/2019/gmd-12-1029-2019.html)

Current version of the module is v.3 and it provides about 150 new diagnostics a new section in the `namelist.input` labelled `&cdxwrf` with new parameters. Outputs are grouped in 3 files:
- `wrfcdx_d<domain>_<date>`: main output file at `auxhist19` with most of the variables
- `wrfhfcdx_d<domain>_<date>`: high-frequency (e,g, 10 minutes) output file at `auxhist18` only for `RAIN` and `COLMAX` implemented to assist the tracking convective systems 
- `wrfpress_d<domain>_<date>`: p-level interpolated variables at `auxhist23` with the additional variables

You can find more details in the [CDXWRF WIKI](https://wiki.cima.fcen.uba.ar/mediawiki/index.php/CDXWRF)

This work was made possible by the participation of the co-authors of the article, talks with a lot of different researchers from a lot of different countries and even other sciences. Thanks to all of them

LluisFB committed:
Adding README file in new create cdxwrf branch to implement the CORDEX-WRF module
L. Fita, J. Polcher, T. M. Giannaros, T. Lorenz, J. Milovac, G. Sofiadis, E. Katragkou,
  and S. Bastin (2019): CORDEX-WRF v1.3: development of a module for the Weather
  Research and Forecasting (WRF) model to support the CORDEX community, Geosci.
  Model Dev., 12, 1029-1066, doi: 10.5194/gmd-12-1029-2019

LluisFB committed:
Adding registry specific for CORDEX-WRF module

LluisFB committed:
Adding diagnostics from CORDEX-WRF module

LluisFB committed:
Adding main subroutines with all CORDEX-WRF diagnostics

LluisFB committed:
Adding modifications for the CORDEX-WRF module

LluisFB committed:
Adding CORDEX-WRF modifications: new p-level variables and mass-conservative vertical interpolation

LluisFB committed:
Initializing CORDEX-WRF diagnostics

LluisFB committed:
Adding CORDEX-WRF water budget related diagnostics

LluisFB committed:
Adding CORDEX-WRF constants

LluisFB committed:
Adding compilation of the CORDEX-WRF module

LluisFB committed:
Adding CORDEX-WRF compilation

LluisFB committed:
Adding high-frequency variables: prhf & colmax on auxhist8

LluisFB committed:
Adding high-frequency CORDEX-WRF output

LluisFB committed:
Adding:
 - `var_colmax1D': Subroutine to compute 1D maximum radar reflectivity

LluisFB committed:
Adding CDXLON, CDXLAT into auxhist8

LluisFB committed:
Getting XLAT, XLONG in auxhist23

LluisFB committed:
Fixing some variables

LluisFB committed:
Fixing some errors comparing with working version

LluisFB committed:
Adding high-frequency CORDEX-WRF diagnostics

LluisFB committed:
Setting up CDXWRF version 3.1, from Eskilstua, Sweden. July 2023
It gets a mix of high-frequency variables, new soil variables, debugging also by time and closest point to a given coordinates

Commit:
Adding into collection the 'registry.cordex' file

LluisFB committed:
Updating README.cordex and moving PRHF to CDXWRF=1

LluisFB committed:
Merge branch 'cdxwrf' of github.com:LluisFB/WRF into cdxwrf

LluisFB committed:
Adding debugging based in time and coordinates

LluisFB committed:
Updating README

LluisFB committed:
Updating README adding the new variables

LluisFB committed:
Updating README

LluisFB committed:
Updating text of README

LluisFB committed:
Adding README.cordex into the README.md

LluisFB committed:
Formatting the README

LluisFB committed:
Fixing README

LluisFB committed:
Adding the p-level interpolated variables

LluisFB committed:
Adding the main URLs

LluisFB committed:
Adding missing cdxhfnsteps

LluisFB committed:
Adding how to inline remove ##CDXWRF[n]##

Commit:
Moving all CDXWRF output to auxhist18 and auxhist19
Including DZ and DZS variables into wrfcdx_d files

Commit:
Fixing p_lev_diags suite (3 varibales where missing)

Commit:
Checking for auxhist18/19_interval values

LluisFB committed:
Correcting README file

LluisFB committed:
Merge branch 'cdxwrf' of github.com:LluisFB/WRF into cdxwrf

LluisFB committed:
Fixing README

Commit:
Adding cd WRF

LluisFB commented:
Please, to make it work, user needs to define pre-compilation flags and modify Registry files. A new file called `README.cordex` provide the detalis:

1. edit the `configure.wrf` and add the line (after the line `-DNETCDF`)
```
                      -DCORDEXDIAG \
```
2. Set up (or not) the pre-compilation variable `CDXWRF` (after the line `-DCORDEXDIAG`)
```
                      -DCDXWRF=[value] \
```
3. Accordingly to the value given to the pre-compilation variable `CDXWRF` one obtains:

- Without adding the variable: all CORDEX 'Core' variables
- `CDXWRF=1` CORDEX 'Tier' variables: clivg, clivh, zmla, [cape/cin/zlfc/plfc/lidx]{min/max/mean}
- `CDXWRF=2` The same as with `CDXWRF=1` and additional variables: ua, va, ws, ta, press, zg, hur, hus, tfog, fogvisbltymin, fogvisbltymax, fogvisbltymean, tdsmin, tdsmax, tdsmean
- `CDXWRF=3` The same as with `CDXWRF=1,2` and additional residence-time variables
- `CDXWRF=4` The same as with `CDXWRF=1,2` and `3` and the Water-Budget relarted ones: wbacdiabh, wbacpw, wbacpw[c/r/s/i/g/h], wbacf, wbacf[c/r/s/i/g/h], wbacz, wbacz[c/r/s/i/g/h], wbacdiabh{l/m/h}, wbacpw{l/m/h}, wbacpw[c/r/s/i/g/h]{l/m/h}, wbacf{l/m/h}, wbacf[c/r/s/i/g/h]{l/m/h}, wbacz{l/m/h}, wbacz[c/r/s/i/g/h]{l/m/h}

4. make a copy of `Registry/registry.cordex`
```
     $ cp Registry/registry.cordex Registry/registry.cordex_comp
```
5. modify the `Registry/registry.cordex_comp` accordingly to the value of `CDXWRF`:
      - Without adding `CDXWRF`, nothing needs to be changed
      - Adding `CDXWRF=1`, one needs to remove the comment `##CDXWRF1##` at the beginning of the line of the definition of certain variables
    - Adding `CDXWRF=2`, one needs to remove the comment `##CDXWRF1##` and `##CDXWRF2##` at the beginning of the line of the definition of certain variables
    - Adding `CDXWRF=3`, one needs to remove the comment `##CDXWRF1##`, `##CDXWRF2##` and `##CDXWRF3##` at the beginning of the line of the definition of certain variables
    - Adding `CDXWRF=4`, one needs to remove the comment `##CDXWRF1##`, `##CDXWRF2##`, `##CDXWRF3##` and `##CDXWRF4##` at the beginning of the line of the definition of certain variables
  For example:
```
sed -i 's/##CDXWRF1##//g' Registry/registry.cordex_comp
sed -i 's/##CDXWRF2##//g' Registry/registry.cordex_comp
sed -i 's/##CDXWRF3##//g' Registry/registry.cordex_comp
```

weiwangncar commented:
@LluisFB Can you compile the code without making any of the changes in configure file and registry? This would be the first step so that the addition of your proposed code would not change the default compilation and running. As we are not receiving any output from the regression tests, it may be an indication that something is not quite right in your code.

weiwangncar committed:
Remove an empty line in solve_em.F to trigger a test.

weiwangncar commented:
The regression test results:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

LluisFB commented:
Yes I am aware that there might be too much pre-compilation flags.

There are about 200 variables with the module. From out analysis (published in [GMD](https://www.geosci-model-dev.net/12/1029/2019/gmd-12-1029-2019.html)), we realized that the less variables you move around, the faster the model runs, also we can easily imagine that not all the people want all the variables.

I grouped them accordingly to what I think is more interesting for the users. There are 2 pre-compilation flags. With the first one (<CODE>-DCORDEXDIAG</CODE> and <CODE>-DCDXWRF=1</CODE>) obtains the CORE variables requested by CORDEX. With the <CODE>CDXWRF=[N]</CODE> (up to 4) there are plenty of additional ones that I had been adding. 

As requested by @weiwangncar there is the other [PR #2247](https://github.com/wrf-model/WRF/pull/2247), where the model has all the variables activated (<CODE>-DCDXWRF=4</CODE>).

Maybe it can be reorganized and left it only with 3 flags? Or split all the module in other additional pieces? 

What do you think?

dudhia commented:
We have tried all the CDXWRF levels. 2 and 3 worked.
WRFCDX_1 and_ WRFCDX_All have the same variables causing the compile error. 'press' may only be defined above CDXWRF=2
WRFCDX_4 seems to not be activating this code and DIABH is not showing up for compile. This is the only #elif, so maybe it is related to that.
#ifdef CDXWRF

#if CDXWRF>=1



#elif CDXWRF>=4

                    ,DIABH=grid%h_diabatic                                 &

                    ,QVTTEND=grid%qvttend, QCTTEND=grid%qcttend            & 

                    ,QRTTEND=grid%qrttend, QSTTEND=grid%qsttend            &

Commit:
Creation of a new README and moving staff from README.md

Commit:
Adding sed instruction into the compilation explanation

Commit:
Fix dimension definition

Commit:
Fixing use of pre-compilation flag. It must be CDXWRF>=3 !!

Commit:
Fixing pre-compilation flags for CDXWRF=1

LluisFB commented:
Dear @smileMchen,

My apologies. There was an issue in the pre-compilation flags. I just committed the fix.

I should checked all the combinations.

Sorry,

Lluís

weiwangncar commented:
@LluisFB Have you tried to compile the modified code on your system? The regression tests returned some syntax errors in module_diag_cordex.F. See one of the test output files here.
[output_3.txt](https://github.com/user-attachments/files/24600853/output_3.txt)

Commit:
Adding complete installation steps from Lluis Fita's WRF fork

weiwangncar commented:
@LluisFB I cloned your code and checked out cdxwrf branch. I'm not using any of your options. I am only trying to compile default WRF code, without using any of your pre-compilation flags. This is to make sure your modification would not affect default use of the model.

Commit:
Empting module_diag_cordex.F when '-DCORDEXDIAG' is not used

LluisFB commented:
HI @weiwangncar 

I see what happens. The module is compiled even if the pre-compilation flags are not activated. 

I do not know if it is an _'elegant'_ solution, but in my last commit, I emptied the module when the main flag is not activated
```
MODULE module_diag_cordex
#ifdef CORDEXDIAG

(....)
#endif
END MODULE module_diag_cordex
```
My apologies, for all the troubles I am making.I should have made all these tests before submitting the Pull Request. Sorry for making waste your time.

I am re-compiling again with CDXWRF=4 to see if it does crashes the compilation

Lluís

weiwangncar commented:
@LluisFB Will compilation work if you do not use any CDXWRF pre-compilation flags?

LluisFB commented:
Yes @weiwangncar, I compiled as it is without using any flag

LluisFB commented:
Well, good news! 

Again apologies for all the foreseeable issues.

Lluís

smileMchen commented:
> Hi @smileMchen,
> 
> Did you tried the last version? With all the new commits? I will re-check for CDXWRF=1

I used the 2nd latest commit. I haven't tried the latest one.

islas commented on Registry/registry.diags:
```diff
@@ -30,12 +46,35 @@ state    real   rh_pl  i{np}j   misc    1  Z   h{23}  "RH_PL"   "Pressure level
 state    real   s_pl   i{np}j   misc    1  Z   h{23}  "S_PL"    "Pressure level data, Speed"                 "m s-1"
 state    real   td_pl  i{np}j   misc    1  Z   h{23}  "TD_PL"   "Pressure level data, Dew point temperature" "K"
-state    real   q_pl   i{np}j   misc    1  Z   h{23}  "Q_PL"    "Pressure level data, Mixing ratio"          "kg/kg"
+state    real   q_pl   i{np}j   misc    1  Z   h{23}  "Q_PL"    "Pressure level data, water vapour Mixing ratio"          "kg/kg"
```
Was changing the description intentional?

islas commented on phys/module_diag_cordex.F:
This file could instead be conditionally be compiled (see frame/Makefile RSL_OBJS for an example on control of this via a variable). Likewise, the compile logic and file additions should be equally done in the CMake build as well.

islas replied:
@LluisFB, no worries. I sincerely believe that this sort of discussion is useful in fostering a collaborative community and finding how to best incorporate your changes into the baseline.

For 1. & 2., my suggested compile changes should not affect the resultant logic downstream. 

I will go ahead and make the edits as soon as possible. This should also give all of us a chance to review the final changes.

Commit:
Recovering empty line

Commit:
Fixing number of auxhist for "wrfhfcdx_d<domain>_<date>" is 18, for "wrfcdx_d<domain>_<date>" is 19

LluisFB commented:
Here I provide some answers to all the points.

First I would like to clarify, that I had been using the module mainly with `CDXWRF=3`, I am aware that I should, (but I haven't) tested all the options before I created the PR. My most sincere apologies.

Now I am running all of them.

- wrfcdx has nothing in it for level 1.
README.cordex was not updated. In a previous version `auxhist` where 8 and 9 for the new outputs. I modified them to 18 and 19, because they collide with another suit. I actualized `README.cordex` accordingly

- Output history 8 was designated to wrfhfcdx_d01 but is never filled.
I fixed it with my previous commit

- What is the meaning of the bins in some of the level 4 output?
These variables were introduced during the creation of the v2. I made a [pdf](https://wiki.cima.fcen.uba.ar/images/8/85/CDXWRF_version2.pdf) where more details are explained.
This is coming from the _residence time_ variables. These variables are designed to provide a 2-dimensional histogram of the time passed simultaneously during the simulation in a given range of 2 variables at each grid point. Each range `(var1_i, var1_i+1)` defines a bin `var1_bin_i`. The module, at each time step, determines at which bin of var1 and var2 lays and adds the length of the time step into the 2-dimensional array of bins as:

$$residence\underline{ }time_{var1var2}(var1\underline{ }bin_{i},var2\underline{ }bin_{j}) = 
  residence\underline{ }time_{var1var2}(var1\underline{ }bin_{i},var2\underline{ }bin_{j}) + time\underline{ }step$$

There are 3 _residence time_ variables:
- `TASHURSTRESHIGHRES`: high-resolution of simultaneous temporal residence of 2-m temperature and relative humidity
- `TASHURSTRESLOWRES`: low-resolution of simultaneous temporal residence of 2-m temperature and relative humidity
- `WBDSWSSTRES`: simultaneous temporal residence of 10-m earth rotated blowing wind direction (from where it bklows)

`TASHURS` variables attempt to be useful for biology, ecosystems and food production, since it provides similar information as the 'growing degree-day ([GDD)](https://en.wikipedia.org/wiki/Growing_degree-day)'
`WBDSWSSTRES` variable will allow to generate the Wind Rose from the values during integration. There is an option that even allows to include the 'wind-gust' estimation into the Wind-Rose

The variables with the bins provide information about the ranges from which the `residence_time` variables are computed.

- Further some fields like TASHURST* are mostly zero but large 3d arrays.
WRF does not allow to provide 4-dimensional (without time) variables. I tried to implemented it, but I did not succeed. Because `TASHURST` variable is a 2-dimensional histogram, I compressed the variable as follows:
```
TASHURSTHIGHRES(west_East,south_north,nhtasrng,nhhursrng,Time) --> TASHURSTHIGHRES(west_East,south_north,nhtashursrng,Time)

TASHURSTHIGHRES(i,j,ihtas,jhhurs,it) --> TASHURSTHIGHRES(i,j,ijhtashurs,it)
```
where 
```
ihtas = INT(ijhtashurs / nhhurs) + 1
jhhurs = ijhtashurs - nhhurs*ihtas
```
So, the residence time variables become very large, because the third dimension encapsulates 2 dimensions.
- nhtasrng = 38, nhhursrng = 25, nhtashursrng = 950
- nltasrng = 21, nlhursrng = 25, nltashursrng = 525
- nwssrng = 22, nwbdsrng= 33, nwbdswssrng = 726

There are 2 versions for `TASHURST` in order to reduce the size of the variable
 - LOWRES: encompasses a larger range of temperatures, but with coarser bins
 - HIGHRES: encompasses a smaller range of temperatures around the _life forms_, but with finer bins

LluisFB commented:
You welcome. 

I have to work in the [WIKI](https://wiki.cima.fcen.uba.ar/index.php/CDXWRF) page to provide individual information for each variable.

Yes time-residence variables are reset to 0 at each output time. So, user obtains accumulations only between output time-steps.

In order to output graphically the variable, one needs to decompress the 3rd dimension, and decide for which period one wants to get the information. Here I provide some examples
|tashurs high res | tashurs low res | wbds wss res |
|----|----|----|
| <img width="252" height="206" alt="highrestime_full" src="https://github.com/user-attachments/assets/e445dfb0-397a-402a-9cb1-266000b11778" />| <img width="252" height="206" alt="lowrestime_full" src="https://github.com/user-attachments/assets/9a8ab6e1-02a8-46a0-83b2-691bf5ffeffb" /> |<img width="252" height="192" alt="CDXWRF_wbds-wss-restime" src="https://github.com/user-attachments/assets/2f251ef2-997d-4f91-9e31-94c53b3b4820" /> |

LluisFB commented:
I run all the possible configurations NO CDXWRF, CDXWRF=1, CDXWRF=2, CDXWRF=3 and CXWRF=4 and all of them run without issues

islas commented:
@LluisFB I cannot get no `CDXWRF` to compile. I have `-DCORDEXDIAG -DINSTVALS` but it is failing withing `module_diag_cordex.F:~974` :
```
        ! air-pressure
        CALL var_press(p(i,:,j), pb(i,:,j), dimz, press(i,:,j))
```

It seems that `press` is only declared if `CDXWRF` is set. Did I misunderstand the instructions? As I read them, `-DCDXWRF` does not need to be set at all within the defines:
> 6.- Set up (or not) the pre-compilation variable CDXWRF (after the line -DCORDEXDIAG)
>                      -DCDXWRF=[value] \
>
> Accordingly to the value given to the pre-compilation variable CDXWRF one obtains:
>   - Without adding the variable: all CORDEX 'Core' variables

Do you mean that `CDXWRF` does not neet to be set _to a value_? Or perhaps, I do not have your latest code..

dudhia commented:
I think @smileMchen (Ming) has compiled this and can help.

LluisFB commented:
Dear @Islas, 

I did not check the `INSTVALS` option. I left it there _'just in case'_ any user wants the instantaneous values of certain diagnostics. I will have to go throughout the code to get it.

I will try do have a look into it as soon as possible

dudhia commented:
@islas Did you try without INSTVALS? I think @smileMchen tested without and it was OK.

islas commented:
@LluisFB @dudhia The error and lines that are causing the issue are unrelated to `INSTVALS`. It really does seem like only using `-DCORDEXDIAG` is not possible - some clarification on the instructions if `-DCDXWRF` (regardless of value) is necessary is needed.

dudhia commented:
To summarize the changes @islas is doing as I understand it
1. To automate the registry contents based on ifdef's rather than require editing
2. To stub out cordex code when the option is not selected with its -D flags.

Commit:
User must provide a level for CDXWRF, otherwise it does not compile

LluisFB commented:
@islas I see and checked in my compilations. 

I did not tried to only set `-DCORDEXDIAG`, without any `-DCDXWRF=[Num]`

I just accordingly modified `README.cordex`

islas committed:
Adjust compilation of CDXWRF to not require user modification of files

islas committed:
Add cordex to cmake

islas reviewed (dismissed):
Ok the compilation has been simplified to use `-cordex N` for the `./configure` script and `-DCDXWRF=N` for `./configure_new`

The two added READMEs still need to be moved to `doc/`. Additionally I don't know if you want to drop one of them since they have similar content

Commit:
Fixing 'frequency' name

Commit:
Removing unncessary readme

Commit:
Moving README to the 'Doc' folder

LluisFB commented:
Thank you very much for all your efforts @islas !

I would appreciate if you can provide some documentation in how it has been made. It will help for any other new introduction of code.

In my last commit I moved `README.cordex` to `doc/` folder and removed unnecessary `README.LluisFB`

islas commented:
> In my last commit I moved `README.cordex` to `doc/` folder and removed unnecessary `README.LluisFB`
@LluisFB I don't see the `doc/README.cordex` file. Did you forget to add it to your commit?

LluisFB commented:
@Islas, I have it there:
``` $ git remote -v 
origin	git@github.com:LluisFB/WRF.git (fetch)
origin	git@github.com:LluisFB/WRF.git (push)

$ git branch
* cdxwrf
  master

$ ls doc
INSTRUCTIONS        README.DA_chem            README.NSSLmp
README.adios2       README.hybrid_vert_coord  README.rsl_output
README.cmake_build  README.hydro              README.SSIB
README.cordex       README.io_config          README.test_cases
README.crtm         README.irr_diag           README.windturbine
README.CTSM         README.madwrf             README.WRFPLUS
README.cygwin.md    README.netcdf4par
README.DA

$ git log -1
commit f612ed082c3d0ed63cc8626c3215957848dbf495 (HEAD -> cdxwrf, origin/cdxwrf)
Author: Lluis Fita Borrell <lluis@debian-BULLSEYE-live-builder-AMD64>
Date:   Wed Feb 11 08:57:12 2026 -0300

    Moving README to the 'Doc' folder
```

islas commented:
Can you do `git status`? It most likely is in your directories but untracked.

Commit:
Moving the file to the right place

LluisFB commented:
Dear @Islas, 

You are right. I moved it, but I forgot to commit in the new location. I thought it was already done, when I commited the `git mv` instrcution. But as I see, it was not the case.

Apologies

islas commented:
@dudhia @weiwangncar I think this is ready for final review now

weiwangncar commented on Registry/registry.diags:
```diff
@@ -30,12 +46,35 @@ state    real   rh_pl  i{np}j   misc    1  Z   h{23}  "RH_PL"   "Pressure level
+state    real   qs_pl_mc i{np}j  misc   1  Z   h{23}  "QS_PL_MC" "mass-conservative pressure level data, snow mixing ratio"  "kg/kg"
+state    real   qi_pl_mc i{np}j  misc   1  Z   h{23}  "QI_PL_MC" "mass-conservative pressure level data, ice mixing ratio"   "kg/kg"
+state    real   qg_pl_mc i{np}j  misc   1  Z   h{23}  "QG_PL_MC" "mass-conservative pressure level data, graupel mixing ratio" "kg/kg"
+state    real   qh_pl_mc i{np}j  misc   1  Z   h{23}  "QH_PL_MC" "mass-conservative pressure level data, hail mixing ratio"  "kg/kg"
```
@LluisFB @islas Will these variables be zero if CDXWRF is not compiled? If so, it needs to be packaged by the CDXWRF namelist option.

LluisFB replied:
This modification has not been done. Now I am not sure how it needs to be done.

@islas who recoded the pre-compilation flags in another way should know

weiwangncar commented on README.md:
```diff
@@ -8,4 +8,3 @@ Information regarding WRF Model citations (including a DOI) can be found here: [
 
 The WRF Model is open-source code in the public domain, and its use is unrestricted. The name "WRF", however, is a registered trademark of the University Corporation for Atmospheric Research. The WRF public domain notice and related information may be found here: [https://www2.mmm.ucar.edu/wrf/users/public.html](https://www2.mmm.ucar.edu/wrf/users/public.html).
 
-
```
@LluisFB No need to make this change.

weiwangncar commented on doc/README.cordex:
```diff
@@ -0,0 +1,372 @@
+ * Pressure interplation
+ - Interpolation on pressure levels of specific humidty (hus_pl), vertical wind speed (w_pl), Earth-rotated wind x-component (uer), Earth-rotated wind y-component (ver) and wind speed (ws) have been also introduced into the `wrfpress' output
+
+* INSTALLATION
```
@LluisFB @islas Are the installation instruction up-to-date with the latest changes in this PR?

LluisFB replied:
No,

`doc/README.cordex` , stil explains the method following my fork in gitHUB.

Now I do not know, how it has to be compiled. @islas should now

islas replied:
I will update them after we cut the release branch

dudhia reviewed (dismissed):
Note it has conflicts. I think we need to "accept both" in the 2 conflicts but it would result in an extra ), so I left it for someone else.

islas committed:
Merge branch 'develop' into cdxwrf

dudhia reviewed (approved):
Check if instructions are up to date with new method.
